rclpy/rclpy/qos.py

```diff
@@ -73,96 +87,100 @@ class QoSProfile:
         '_avoid_ros_namespace_conventions',
     ]
 
-    def __init__(self, **kwargs):
-        assert all('_' + key in self.__slots__ for key in kwargs.keys()), \
-            'Invalid arguments passed to constructor: %r' % kwargs.keys()
+    def __init__(self, history: Optional['QoSHistoryPolicy'] = None, depth: Optional[int] = None,
+                 reliability: Optional['QoSReliabilityPolicy'] = None,
+                 durability: Optional['QoSDurabilityPolicy'] = None,
+                 lifespan: Optional[Duration] = None, deadline: Optional[Duration] = None,
+                 liveliness: Optional['QoSLivelinessPolicy'] = None,
+                 liveliness_lease_duration: Optional[Duration] = None,
+                 avoid_ros_namespace_conventions: Optional[bool] = None) -> None:
 
-        if 'history' not in kwargs:
-            if 'depth' not in kwargs:
+        if history is None:
+            if depth is None:
                 raise InvalidQoSProfileException('History and/or depth settings are required.')
-            kwargs['history'] = QoSHistoryPolicy.KEEP_LAST
+            history = QoSHistoryPolicy.KEEP_LAST
 
-        self.history = kwargs.get('history')
+        self.history = history
 
         if (
-            QoSHistoryPolicy.KEEP_LAST == self.history and
-            'depth' not in kwargs
+            QoSHistoryPolicy.KEEP_LAST == self.history and depth is None
         ):
             raise InvalidQoSProfileException('History set to KEEP_LAST without a depth setting.')
 
-        self.depth = kwargs.get('depth', QoSProfile.__qos_profile_default_dict['depth'])
-        self.reliability = kwargs.get(
-            'reliability', QoSProfile.__qos_profile_default_dict['reliability'])
-        self.durability = kwargs.get(
-            'durability', QoSProfile.__qos_profile_default_dict['durability'])
-        self.lifespan = kwargs.get('lifespan', QoSProfile.__qos_profile_default_dict['lifespan'])
-        self.deadline = kwargs.get('deadline', QoSProfile.__qos_profile_default_dict['deadline'])
-        self.liveliness = kwargs.get(
-            'liveliness', QoSProfile.__qos_profile_default_dict['liveliness'])
-        self.liveliness_lease_duration = kwargs.get(
-            'liveliness_lease_duration',
-            QoSProfile.__qos_profile_default_dict['liveliness_lease_duration'])
-        self.avoid_ros_namespace_conventions = kwargs.get(
-            'avoid_ros_namespace_conventions',
-            QoSProfile.__qos_profile_default_dict['avoid_ros_namespace_conventions'])
+        if depth is None:
+            self.depth = QoSProfile.__qos_profile_default_dict['depth']
+        else:
+            self.depth = depth
+
+        self.reliability = reliability or QoSProfile.__qos_profile_default_dict['reliability']
```

Comment:
I think the `or` logic makes it impossible to use the `SYSTEM_DEFAULT` QoS policy, which IIRC is the one that allows the QoS to be set by middleware specific configuration.

```python3
>>> from enum import IntEnum
>>> class ReliabilityPolicy(IntEnum):
...     SYSTEM_DEFAULT = 0
...     RELIABLE = 1
...     BEST_EFFORT = 2
...     UNKNOWN = 3
...     BEST_AVAILABLE = 4
...
>>> ReliabilityPolicy.SYSTEM_DEFAULT or ReliabilityPolicy.RELIABLE
<ReliabilityPolicy.RELIABLE: 1>
```

I'd recommend instead for all of these:

```python3
self.reliability = QoSProfile.__qos_profile_default_dict['reliability'] if reliability is None else reliability
```

Reply:
Will do!